central/platform/reprocessor/reprocessor_impl.go

```diff
@@ -0,0 +1,210 @@
+)
+
+type platformReprocessorImpl struct {
+	alertDatastore      alertDS.DataStore
+	deploymentDatastore deploymentDS.DataStore
+	platformMatcher     platformmatcher.PlatformMatcher
+
+	stopSignal concurrency.Signal
+	// isStarted will make sure only one reprocessing routine runs for an instance of reprocessor
+	isStarted atomic.Bool
+}
+
+func New(alertDatastore alertDS.DataStore,
+	deploymentDatastore deploymentDS.DataStore,
+	platformMatcher platformmatcher.PlatformMatcher) PlatformReprocessor {
+
+	return &platformReprocessorImpl{
+		alertDatastore:      alertDatastore,
+		deploymentDatastore: deploymentDatastore,
+		platformMatcher:     platformMatcher,
+		stopSignal:          concurrency.NewSignal(),
+	}
+}
+
+func (pr *platformReprocessorImpl) Start() {
+	swapped := pr.isStarted.CompareAndSwap(false, true)
+	if !swapped {
+		log.Error("Platform reprocessor was already started")
+		return
+	}
+	go pr.runReprocessing()
+}
+
+func (pr *platformReprocessorImpl) Stop() {
+	if !pr.isStarted.Load() {
+		log.Error("Platform reprocessor not started")
+	}
+	pr.stopSignal.Signal()
+}
+
+func (pr *platformReprocessorImpl) runReprocessing() {
+	err := pr.reprocessAlerts()
+	if err != nil {
+		log.Errorf("Error reprocessing alerts with platform rules: %v", err)
+	}
+
+	err = pr.reprocessDeployments()
+	if err != nil {
+		log.Errorf("Error reprocessing deployments with platform rules: %v", err)
+	}
+}
+
+func (pr *platformReprocessorImpl) alertsNeedReprocessing() (bool, error) {
+	// Check if there is atleast one alert where platform component is unset
+	q := unsetPlatformComponentQuery.CloneVT()
+	q.Pagination = &v1.QueryPagination{
+		Limit: 1,
+	}
+	alerts, err := pr.alertDatastore.GetByQuery(reprocessorCtx, q)
+	if err != nil {
+		return false, err
+	}
+	return len(alerts) > 0, nil
+}
+
+func (pr *platformReprocessorImpl) deploymentsNeedReprocessing() (bool, error) {
+	// Check if there is atleast one deployment where platform component is unset
+	q := unsetPlatformComponentQuery.CloneVT()
+	q.Pagination = &v1.QueryPagination{
+		Limit: 1,
+	}
+	deployments, err := pr.deploymentDatastore.SearchRawDeployments(reprocessorCtx, q)
+	if err != nil {
+		return false, err
+	}
+	return len(deployments) > 0, nil
+}
+
+func (pr *platformReprocessorImpl) reprocessAlerts() error {
+	if pr.stopSignal.IsDone() {
+		log.Debug("Stop called, stopping platform reprocessor")
+		return nil
+	}
+	needReprocessing, err := pr.alertsNeedReprocessing()
+	if err != nil {
+		return errors.Wrap(err, "Error determining if alerts need reporcessing")
+	}
+	if !needReprocessing {
+		log.Debug("Alerts up to date with platform rules, skipping reprocessing")
+		return nil
+	}
+
+	q := unsetPlatformComponentQuery.CloneVT()
+	q.Pagination = &v1.QueryPagination{
+		Limit:  batchSize,
+		Offset: 0,
+		SortOptions: []*v1.QuerySortOption{
+			{
+				Field: search.AlertID.String(),
+			},
+		},
+	}
```

Comment:
@charmik-redhat It just hit me that I'm not sure this works.  You are getting a page at a time, but also updating them as you get them.  So I don't think the sort nor the offset is necessary.  I think you only need to get 0..batchSize each time.  Because once you update the first batch the paging and offset will be off.  Please test this to verify my statement and address as necessary.

Reply:
Oh yeah great catch. The updated alerts would not match the query criteria anymore. So the next batch should have an offset 0, not batchSize. I will reduce the batchSize to 10 or something to test this out.